File: python/39_Combination_Sum.py
class Solution:
    def backtrack(self, candidates: List[int], target: int, index:int, curPath:List[int], res: List[List[int]] ):
        if(target == 0):
            res.append(curPath[:])
            return
        if(target < 0 or index == len(candidates)):
            return
        
        for i in range(index, len(candidates)):
            if target-candidates[i] < 0:
                break
            
            curPath.append(candidates[i])
            self.backtrack(candidates,target-candidates[i],i,curPath,res)
            curPath.pop()

    def combinationSum1(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        self.backtrack(candidates,target,0,[],res)
        return res

        def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        n = len(candidates)
        res = []
        def helper(index, curSum, curPath):
            if curSum > target or index == n:
                return
            if curSum == target:
                res.append(curPath[:])
                return
            for i in range(index, n):
                if curSum + candidates[index] > target:
                    break
                helper(index, curSum + candidates[index] ,curPath+[candidates[index]])
            
        helper(0,0,[])
        return res

File: java/39_Combination_Sum.java
class Solution {

    public List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        int n = candidates.length;
        dfs(candidates , target , 0, 0 , new ArrayList<Integer>());
        return ans;

    }

    public void dfs(int[] candidates, int target, int index,int curSum, List<Integer> curPath){
        if(curSum == target){
            ans.add(new ArrayList<Integer>(curPath));
            return;
        }
        if(curSum > target){
            return;
        }
        
        for(int i = index; i < candidates.length ; i++){
            curPath.add(candidates[i]);
            dfs(candidates , target , i, curSum + candidates[i] , curPath);
            curPath.remove(curPath.size()-1);
        }    

    }
}